src/core_plugins/i18n/server/i18n/i18n.js

```diff
@@ -0,0 +1,78 @@
+import fs from 'fs';
+import Promise from 'bluebird';
+
+const readFile = Promise.promisify(fs.readFile);
+
+const TRANSLATION_FILE_EXTENSION = 'json';
+
+let registeredTranslations = {};
+
+const getRegisteredTranslationLocales = function () {
+  return Object.keys(registeredTranslations);
+};
+
+const registerTranslations = function (absolutePluginTranslationFilePath) {
+  let translationFiles = [];
+
+  const locale = getLocaleFromFileName(absolutePluginTranslationFilePath);
+  if (registeredTranslations.hasOwnProperty(locale)) {
+    translationFiles = registeredTranslations[locale];
+  }
+  translationFiles.push(absolutePluginTranslationFilePath);
+  registeredTranslations[locale] = translationFiles;
+};
+
+const getRegisteredLocaleTranslations = function (locale) {
+  if (!registeredTranslations.hasOwnProperty(locale)) {
+    throw new Error('No translations registered for locale: ' + locale);
+  }
+
+  const translationFiles = registeredTranslations[locale];
+  let translationJson = {};
+
+  return Promise.map(translationFiles, (translationFile) => {
+    return readFile(translationFile, 'utf8').then(function (translationStr) {
+      const translationToAddJson = JSON.parse(translationStr);
+      for (let key in translationToAddJson) {
+        if (translationToAddJson.hasOwnProperty(key)) {
+          const attrName = key;
+          const attrValue = translationToAddJson[key];
+          translationJson[attrName] = attrValue;
+        }
+      }
+    });
+  }).then(function () {
+    return translationJson;
+  });
```

Comment:
Rewritten this code block to take your comments on board. I agree that it has a better flow to it. I have also added error handling and verification of the json schema.